fix: avoid stopping propagation when not necessary #11457

ncoden opened the pull request:
<!--- --------------------------------------------------------------------- -->
<!---                 Please fill the following template                    -->
<!---             Your pull request may be ignored otherwise                -->
<!--- --------------------------------------------------------------------- -->

## Description
<!--- Describe your changes in detail. Include any relevant information     -->
<!--- (reasons, difficulties, links to web references, related issues...)   -->

Stopping the event propagation prevent the user to watch for the event at an higher level. If it was needed to prevent a duplicate behavior in a component, others methods should be used like debouncing the event.

This pull request remove `stopPropagation` and `stopImmediatePropagation` from everywhere we do not actually want to "cancel" the event.

### Changes:
- Save and check for `event.zfIsKeyHandled` in Keyboard utility to prevent to handle twice the same event in different components.
- Remove `stopPropagation` from components' keyboard handlers (Accordion, DropdownMenu, OffCanvas and Tabs).
- Remove `stopPropagation` and `stopImmediatePropagation` from everywhere we do not want to cancel the event (AccordionMenu, Drilldown, DropdownMenu, Tabs, Tooltip).
- Update `DropdownMenu._removeBodyHandler` to also ignore click events on the element itself.

### Others changes:
- Clean up `DropdownMenu._events` a bit.
- Add explainaitions where `stopPropagation` is still used (OffCanvas, Triggers).


### Actions taken to avoid regression:
- [x] Check for when `stopPropagation` was added, the reasons behind it and if it resolved any bugs (#8691).
- [x] Search for any related issues.
- [x] Run all visual tests.

### Related
- See [The Dangers of Stopping Event Propagation](https://css-tricks.com/dangers-stopping-event-propagation/) (@philipwalton)
- Closes #11457

## Types of changes
<!--- What types of changes does your code introduce?                       -->
<!--- Put an `x` in all the boxes that apply:                               -->
- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to change)

## Checklist (all required):
<!--- Go over all the following points, and put an `x` in the boxes.        -->
<!--- If you're unsure about any of these, don't hesitate to ask.           -->
- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] There are no other pull request similar to this one.
- [x] The pull request title is descriptive.
- [x] The template is fully and correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [x] My code follows the code style of this project.
- ~~I have updated the documentation accordingly to my changes (if relevant).~~
- ~~I have added tests to cover my changes (if relevant).~~
- [x] All new and existing tests passed.

<!--- --------------------------------------------------------------------- -->
<!---       For more information, see the CONTRIBUTING.md document          -->
<!---         Thank you for your pull request and happy coding ;)           -->
<!--- --------------------------------------------------------------------- -->

ncoden committed:
docs: add docs about "mobile scroll" bug #9707 in OffCanvas

Add documentation for the OffCanvas private methods `_recordScrollable` and `_stopScrollPropagation` to explicitely justify the need for `stopPropagation`.

See #9707

ncoden committed:
fix: prevent double Keyboard events handling without stopping propagation

Stopping the event propagation prevent the user to watch for the event at an higher level. If it was needed to prevent a duplicate behavior in a component, others methods should be used like debouncing the event.

Changes:
- Save and check for `event.zfIsKeyHandled` in Keyboard utility to prevent to handle twice the same event in different components.
- Remove `event.stopPropagation` from components' keyboard handlers (Accordion, DropdownMenu, OffCanvas and Tabs).

See https://github.com/zurb/foundation-sites/issues/11457

ncoden committed:
fix: avoid stopping event propagation when not necessary

Stopping the event propagation prevent the user to watch for the event at an higher level. If it was needed to prevent a duplicate behavior in a component, others methods should be used like debouncing the event.

Changes:
- Remove `stopPropagation` and `stopImmediatePropagation` from everywhere we do not want to cancel the event (AccordionMenu, Drilldown, DropdownMenu, Tabs, Tooltip).
- Update `DropdownMenu._removeBodyHandler` to also ignore click events on the element itself.

Others changes:
- Clean up `DropdownMenu._events` a bit.
- Add explainaitions where `stopPropagation` is still used (Triggers).

See https://github.com/zurb/foundation-sites/issues/11457

SassNinja commented on js/foundation.offcanvas.js:
```diff
@@ -296,13 +305,22 @@ class OffCanvas extends Plugin {
 
     if((up && elem.allowUp) || (down && elem.allowDown)) {
+      // It is not recommended to stop event propagation (the user cannot watch it),
+      // but in this case this is the only solution we have.
```
well spotted 👍 
(`contentScroll: false` does not work correctly on iOS without this)

ncoden replied:
Yes, I gone through the history of every `stopPropagation` and added comments when they were needed.

SassNinja commented on js/foundation.dropdownMenu.js:
```diff
@@ -286,8 +284,8 @@ class DropdownMenu extends Plugin {
     $body.on('click.zf.dropdownMenu tap.zf.dropdownMenu', (e) => {
-      var $link = this.$element.find(e.target);
-      if ($link.length) return;
+      var isItself = this.$element.andSelf().find(e.target).length;
```
`andSelf()` is deprecated and should be replaced with `addBack()`

Apart from that I think the order is wrong. When you want to include the elem itself it should be
```javascript
var isItself = this.$element.find(e.target).addBack().length;
```

ncoden replied:
I used `$(e.target).closest(this.$element).length`. I think that testing the DOM tree upward is way faster than browsing all the nodes downward.

SassNinja replied:
Good idea 👍

ncoden committed:
fix: fix DropdownMenu body handler check to ignore the element itself

> `andSelf()` is deprecated and should be replaced with `addBack()`
> -- @SassNinja

Also, `andSelf` cannot be used that way. This commit replace it with the opposite approach: checking if the target is the child of the element or the element itself using `closest()`.

SassNinja reviewed (approved):
LGTM

ncoden commented:
Thank you for the review @SassNinja

ncoden commented:
Targeted for `v6.6.0` because this may break things depending on the event listeners in the user project.

joeworkman commented:
This pull request has been mentioned on **Foundation Open Source Community**. There might be relevant details there:

https://foundation.discourse.group/t/foundation-v6-6-1-v6-6-0-breaks-nested-dropdown-menus/1012/3

DanielRuf commented:
@SassNinja there is a case where it breaks setups. Can you check https://foundation.discourse.group/t/foundation-v6-6-1-v6-6-0-breaks-nested-dropdown-menus/1012?
